[ENT-303] Allow coupon API to be filtered by EnterpriseCustomer UUID; allow catalog-based coupons to be linked to EnterpriseCustomers

haikuginger opened the pull request:
**Description:**

This pull request adds two feature:

First, the ability to link a catalog-based coupon to an EnterpriseCustomer. We had initially disallowed this, but it's become apparent that this is a necessary use case, so this pull request removes the Javascript that disallowed it.

Second, the ability to filter, in the API, coupons by the EnterpriseCustomer that they're linked to. This is accomplished by passing an `enterprise_customer` query parameter containing the UUID of the EnterpriseCustomer in question.

**JIRA:** [ENT-303](https://openedx.atlassian.net/browse/ENT-303)

**Dependencies:** None

**Merge deadline:** 24 April 2017

**Installation instructions:** Standard ecommerce installation

**Testing instructions:**

1. In the coupon creation view, verify that you can create a coupon tied to a catalog that is also tied to an EnterpriseCustomer.
2. In the coupon list API view, verify that adding an `enterprise_customer` query parameter that is a valid UUID filters the coupons down to only those coupons that are linked to that EnterpriseCustomer.

edx-webhook commented:
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=1247"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

vkaracic commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -53,6 +54,28 @@ def get_serializer_class(self):
+        Handle cases where we need to filter by fields that aren't
+        natively supported on the Coupon object.
+        """
+        ec_filter = self.request.GET.get('enterprise_customer', None)
```
You don't need ``None`` here.

haikuginger replied:
Right you are!

vkaracic commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -53,6 +54,28 @@ def get_serializer_class(self):
+        """
+        ec_filter = self.request.GET.get('enterprise_customer', None)
+        objs = super(CouponViewSet, self).get_queryset()
+        if ec_filter is not None:
```
Can you just say ``if ec_filter:``?

haikuginger replied:
No; that'd match blank strings, which aren't valid filters. To be fair, that'd get caught by checking if it's a valid UUID, but I prefer to be explicit.

clintonb replied:
Filtering should be done in `filters.py` with `django_filters. See https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/api/filters.py.

vkaracic commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -480,6 +480,35 @@ def test_list_coupons(self):
+        self.assertEqual(coupon_data['client'], self.data['client'])
+        response = self.client.get(COUPONS_LINK)
+        # And test that without filtering, we get both coupons.
+        self.assertEqual(len(json.loads(response.content)['results']), 2)
```
You should add a few blank lines to make the code more legible.

haikuginger replied:
Agreed! Legible code is always good.

vkaracic commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -480,6 +480,35 @@ def test_list_coupons(self):
+        coupon_data = json.loads(response.content)['results'][0]
+        self.assertEqual(coupon_data['title'], self.data['title'])
+        self.assertEqual(coupon_data['category']['name'], self.data['category']['name'])
+        self.assertEqual(coupon_data['client'], self.data['client'])
```
If I understand this test correctly you get the same thing with a fake UUID as you do with a valid one?

haikuginger replied:
No. Initially, we have one coupon, and so calling the list API will return that single coupon. When calling with an invalid UUID, we get that single coupon back, since an invalid UUID can't filter anything. We then add an _additional_ coupon that has a UUID attached, and attempt to filter by that valid UUID, resulting in a single (different) result.

itsjeyd commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -480,6 +480,35 @@ def test_list_coupons(self):
+        self.assertEqual(coupon_data['category']['name'], self.data['category']['name'])
+        self.assertEqual(coupon_data['client'], self.data['client'])
+        response = self.client.get(COUPONS_LINK)
+        # And test that without filtering, we get both coupons.
```
@haikuginger I think we should move this comment above the `GET` in the previous line.

haikuginger replied:
Sure!

itsjeyd commented on ecommerce/extensions/api/v2/tests/views/test_coupons.py:
```diff
@@ -480,6 +480,35 @@ def test_list_coupons(self):
+        coupon_data = json.loads(response.content)['results'][0]
+        self.assertEqual(coupon_data['title'], self.data['title'])
+        self.assertEqual(coupon_data['category']['name'], self.data['category']['name'])
+        self.assertEqual(coupon_data['client'], self.data['client'])
```
@haikuginger Since these assertions are duplicated below it might make sense to move them to a separate method, something like `assert_coupon_data`.

haikuginger replied:
Sure; that makes sense!

itsjeyd commented on ecommerce/static/js/test/specs/views/coupon_form_view_spec.js:
```diff
@@ -201,16 +201,6 @@ define([
-
-                    view.$('#multiple-courses').prop('checked', true).trigger('change');
-                    expect(SpecUtils.formGroup(view, '[name=enterprise_customer]')).toBeVisible();
-                });
```
@haikuginger IIUC, we now want the menu for selecting an enterprise customer to be visible at all times. Would it be worth keeping this test and just changing `.not.toBeVisible()` in L209 to `.toBeVisible()`, to make sure any changes to the current behavior wouldn't go unnoticed?

haikuginger replied:
We definitely could; that's a good idea.

itsjeyd commented on ecommerce/extensions/api/v2/views/coupons.py:
```diff
@@ -53,6 +54,28 @@ def get_serializer_class(self):
+                    coupon_vouchers__vouchers__offers__condition__range__enterprise_customer=ec_filter
+                ).distinct()
+        return objs
+
```
@haikuginger Thanks for adding detailed comments about what's happening to this method :)

itsjeyd commented:
@haikuginger I tested the changes according to the instructions in the PR description and they are working fine. Couldn't associate a coupon with a catalog because ecommerce is currently failing to retrieve course catalog data from the Course Discovery API, but that's not related to the changes that this PR introduces. The "Enterprise Customer" option is available for any of "Single course", "Multiple courses", "Catalog".

haikuginger commented:
@clintonb, I think you're probably right that we should prioritize that work. However, we'd still like to get this across the line. Would this change be acceptable to you if we moved the actual query to filter by EnterpriseCustomer UUID into the `ecommerce.enterprise` module?

Doing so would isolate where we need to make changes during the refactor to just that module. Otherwise, the needed behavior should remain the same - we need to get a list of the coupons that are associated with a particular EnterpriseCustomer, whether we do that by doing a database lookup to the related Range, or if we locate an Enterprise-linked Offer and get the linked coupons that way.

haikuginger commented:
@marjev @vkaracic @itsjeyd, I've addressed your comments. @clintonb, I haven't made any changes to the filtration logic yet, pending a discussion on if and how we should go ahead with implementing this right now, or wait on #1218.

clintonb commented:
This cannot be merged until #1218 is addressed. Building atop the incorrect model will only make fixing that model more difficult.

haikuginger commented:
@clintonb, okay! I actually just got off a call with Griff, and we've agreed to defer this until later. I'm going to go ahead and close this PR, as it's no longer necessary.

@itsjeyd, @vkaracic, @marjev, thank you for your reviews!
